search/shortest.go

```diff
@@ -0,0 +1,263 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package search
+
+import (
+	"math"
+	"math/rand"
+
+	"github.com/gonum/graph"
+	"github.com/gonum/matrix/mat64"
+)
+
+// Shortest is a shortest-path tree created by the DijkstraFrom function.
+type Shortest struct {
+	// from holds the source node given to
+	// DijkstraFrom.
+	from graph.Node
+
+	// nodes hold the nodes of the analysed
+	// graph.
+	nodes []graph.Node
+	// indexOf contains a mapping between
+	// the id-dense representation of the
+	// graph and the potentially id-sparse
+	// nodes held in nodes.
+	indexOf map[int]int
+
+	// dist and next represent the shortest
+	// paths between nodes.
+	//
+	// Indices into dist and next are
+	// mapped through indexOf.
+	//
+	// dist contains the distances
+	// from the from node for each
+	// node in the graph.
+	dist []float64
+	// next contains the shortest-path
+	// tree of the graph. The index is a
+	// linear mapping of from-dense-id
+	// and to-dense-id.
+	next []int
```

Comment:
Yeah, I struggled with the naming of the equivalent field in the all paths equivalent type (considering calling it step) but settled on next because it contains the next item to look at - rather than the next node in the path. The field here is named to evoke similarity between the types.